File: src/main/scala/dxw405/gui/TextFieldPlaceholder.scala
package dxw405.gui

import javax.swing._
import java.awt._

class TextFieldPlaceholder(placeHolderText:String) extends JTextField
{
	private val placeholder: String = placeHolderText

	override def paintComponent(g: Graphics)
	{
		super.paintComponent(g)
		if (getText.length == 0)
		{
			g.asInstanceOf[Graphics2D].setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON)
			val m: Int = 0xfefefefe
			g.setColor(new Color(((getBackground.getRGB & m) >>> 1) + ((getForeground.getRGB & m) >>> 1), true))
			g.drawString(placeholder, getInsets.left, getHeight / 2 + g.getFontMetrics.getAscent / 2 - 2)
		}
	}
}

File: src/main/java/dxw405/util/TextFieldPlaceholder.java
package dxw405.util;

import javax.swing.*;
import java.awt.*;

public class TextFieldPlaceholder extends JTextField
{
	private String placeholder;

	public TextFieldPlaceholder(String placeholder) {this.placeholder = placeholder;}

	@Override
	public void paintComponent(Graphics g)
	{
		super.paintComponent(g);
		if (getText().length() == 0)
		{
			int h = getHeight();
			((Graphics2D) g).setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON);
			Insets ins = getInsets();
			FontMetrics fm = g.getFontMetrics();
			int c0 = getBackground().getRGB();
			int c1 = getForeground().getRGB();
			int m = 0xfefefefe;
			int c2 = ((c0 & m) >>> 1) + ((c1 & m) >>> 1);
			g.setColor(new Color(c2, true));
			g.drawString(placeholder, ins.left, h / 2 + fm.getAscent() / 2 - 2);
		}
	}
}
